_posts/2019-05-20-server-rendering-responsively.md

```diff
@@ -0,0 +1,201 @@
+server-rendered React app. When using SSR in a React app, you need to call
+[`ReactDOM.hydrate()`](https://reactjs.org/docs/react-dom.html#hydrate) to connect your server-rendered app to the
+client's browser. As the docs describe,
+
+> React expects that the rendered content is identical between the server and the client.
+
+If there are differences between your server-rendered component tree and the initial client-rendered component
+tree, a couple things will happen.
+
+1. You'll get a friendly error in your console telling you there were differences (in development mode), but
+   also...
+2. weird, unpredictable things can happen to your app.
+
+Here's what our app looked like in one case where our client render didn't match our server render:
+
+![A site whose client render does not match its server render, resulting in spinning body copy](/images/2019-05-20-server-rendering-responsively/wacky-spinner.gif)
+
+Yuck! The docs explain that `ReactDOM.hydrate` doesn't put a lot of effort into patching up differences between the
+server and client renders, and that's how you end up with weird things like above. Weird things don't _always_
+happen, but they _can_ happen. Your best bet is to match server and client renders, to avoid strange results that
+are difficult to explain or debug.
+
+Because hydration requires the component trees to match between server and client, we render _all_ breakpoints from
+the server with `react-responsive-media`. This way, we know you're going to get a branch that matches between the
+client and server. Then, upon hydration, we prune the branches that don't match the current breakpoint.
+
+If you're interested, you can
+[read the issue that originally inspired us to build `@artsy/react-responsive-media`](https://github.com/artsy/reaction/issues/1367).
+One of the neat things about Artsy being [open-source by default](/series/open-source-by-default/) is that you can
+see decisions being made and libraries being built _as they happen_; not just after they're complete.
+
+## Tool 3: [`@artsy/detect-responsive-traits`](https://github.com/artsy/detect-responsive-traits)
+
+I mentioned above that it's difficult to accurately detect devices by user agent to identify which breakpoint to
+render. We didn't want this to be our primary strategy for combining SSR with responsive design.
+
+But with `react-responsive-media` as our primary approach, we felt that we could make some further optimizations
+with user agent detection. In the event that we don't know your device by its user agent, we'll still give you both
+the desktop and mobile breakpoints from SSR. But if we are certain you are on a mobile device, we can omit the
+desktop breakpoint from the server render.
+
+We really wanted to not maintain our own list of user agents. Alas, we found that none of the existing user agent
+detection libraries surfaced all the information we needed. We needed to know the minimum width for a browser on a
```

Comment:
```suggestion
detection libraries surfaced all the information we needed in a single resource. We needed to know the minimum width for a browser on a
```

Not sure if that's the best way to explain that we needed to get bits of data from various sources.